### PR TITLE
#345 Application home page

### LINKED ISSUE #345: Implement ApplicationHome

### PR DESCRIPTION
**Pull request by nmadruga**
Fixes #345 

### Branched from Issue's PR:
~~#371~~ MERGED 

### Changes
- New component `ApplicationHeader` to render the top part of the  container on Home page and **back to list** button. It also receives the content to be displayed inside its container
- New component `SectionsProgress`  to display  all sections in the application (using **fullStructure.sections** with each progress on sections and a **Resume** button on the first section with invalid/incomplete page
- Find the correct first section with invalid page to be done in issue #374 
- When all sections are complete show the **Review & Summary** button on the Home page
- When the application is not in `Draft` or `Changes Requested` status redirect to the Summary page (Issue #367 in progress)

### Unrelated changes
- Add `fluid` to main container to display the application a little better on mobile
- Still can't find a simple way to display the progress nicely on mobile. Probably something to do as part of [Epic#10 - App menu & layout](https://github.com/openmsupply/application-manager-web-app/issues/78)
Screen Shot 2021-02-25 at 10 39 56 PM

### CODE REVIEW

File: src/containers/Main/AuthenticatedWrapper.tsx
Comment (CarlosNZ) [inline src/containers/Main/AuthenticatedWrapper.tsx:10]: This makes the app go full-width which looks pretty bad. If we need to use it to be responsive, then maybe add a `maxWidth` for now? We need to have a plan about implementing layout/styles at some point, but for now we should probably keep it like it is for the most part.

Just looked up `Container` in Semantic: "A container limits content to a maximum width."
And then also: "fluid |   | {bool} | Container has no maximum width.'

If all the `fluid` prop is doing is turning off maximum width, why do we need it?

File: src/containers/Application/ApplicationHome.tsx
Comment (CarlosNZ) [inline src/containers/Application/ApplicationHome.tsx:43]: `/summaryNEW`  ?

File: src/containers/Application/ApplicationHome.tsx
Comment (CarlosNZ) [inline src/containers/Application/ApplicationHome.tsx:35]: `/summaryNEW`?

Comment (CarlosNZ) [reply src/containers/Application/ApplicationHome.tsx:35]: Also -- why `replace` rather than `push`?

File: src/containers/Application/ApplicationHome.tsx
Comment (CarlosNZ) [inline src/containers/Application/ApplicationHome.tsx:48]: Can't help thinking that this might be better calculated and stored during `generateProgress`. Do you know if this is required elsewhere? If so, probably better to do it during the hook. If not, doesn't really matter.

File: src/components/Sections/SectionsProgress.tsx
Comment (CarlosNZ) [inline src/components/Sections/SectionsProgress.tsx:34]: You should be able to add this now from using my PR#376

File: src/components/Sections/SectionsProgress.tsx
Comment (CarlosNZ) [inline src/components/Sections/SectionsProgress.tsx:55]: Just wondering about the placement of the button. The way it is now, it looks like it's tied to Section 1, which is a bit misleading. (esp since Section 1 is completed in this case):
Screen Shot 2021-02-26 at 10 38 27 AM

Maybe place it on its own line below the last section?

**Review by CarlosNZ — CHANGES_REQUESTED** [review]
Great. Just some minor comments, although I do think we need to put the Container back as it's looking pretty bad when the browser window is wide.

File: src/containers/Main/AuthenticatedWrapper.tsx
Comment (nmadruga) [reply src/containers/Main/AuthenticatedWrapper.tsx:10]: Good point... I'm removing it now, it doesn't seen to have improved at all.

File: src/containers/Application/ApplicationHome.tsx
Comment (nmadruga) [reply src/containers/Application/ApplicationHome.tsx:43]: No need. Just using `/applicatioNEW/...` is enough to get to the proper new ApplicationSummary page

File: src/containers/Application/ApplicationHome.tsx
Comment (nmadruga) [reply src/containers/Application/ApplicationHome.tsx:35]: I was just keep the same behaviour as before. But if pushing is better (to allow user going back to previous page) I can change to that. Thanks for the suggestion

File: src/containers/Application/ApplicationHome.tsx
Comment (CarlosNZ) [reply src/containers/Application/ApplicationHome.tsx:43]: Ah yes of course.

File: src/containers/Application/ApplicationHome.tsx
Comment (nmadruga) [reply src/containers/Application/ApplicationHome.tsx:48]: Hmmm it would be required in the ApplicationPage as well I think - if we want to display the **Review & Submit** button with different appearance when the application has all required elements completed.

File: src/components/Sections/SectionsProgress.tsx
Comment (nmadruga) [reply src/components/Sections/SectionsProgress.tsx:34]: Great. I'll rather wait for it to be merged into master first - to avoid merging confusion between branches

File: src/components/Sections/SectionsProgress.tsx
Comment (nmadruga) [reply src/components/Sections/SectionsProgress.tsx:55]: It's just showing the button in here because the `firstIncomplete` not filled out correctly yet. So it would display in Section 1 for any application. Until I get the changes on master and update this to use the correct one...

**Comment by nmadruga** [comment]
Continuing [this discussion](https://github.com/openmsupply/application-manager-web-app/pull/375#discussion_r583221116)

> Maybe place it on its own line below the last section?

Although this may be better since it would look nice in both Home pages (Review and Application) since for now they have different UI designs and if we are going to combine and use the same component we need the same  way to display this **Resume** button.

Current UI designs:
Application
Screen Shot 2021-02-26 at 11 56 02 AM

Review
Screen Shot 2021-02-26 at 11 57 39 AM

Although... thinking a little further on this, there is also the case when Applicant receives the application back with "Changes required" or the Review/Consolidator do to, which have these UI designs (see there are links/buttons per sections): 

Application
Screen Shot 2021-02-26 at 11 56 48 AM

Review
Screen Shot 2021-02-26 at 12 01 33 PM

File: src/containers/Application/ApplicationHome.tsx
Comment (CarlosNZ) [reply src/containers/Application/ApplicationHome.tsx:47]: Actually, we don't need to re-calculate this at all. We just need to check `fullStructure.info.firstIncompletePage` -- if it's `null`, then the application is "completed" :)

File: src/components/Application/ApplicationHeader.tsx
Comment (andreievg) [inline src/components/Application/ApplicationHeader.tsx:36]: not sure if we need "strings.TITLE_APPLICATION_FORM}" or "{strings.TITLE_INTRODUCTION}", but I guess can be removed later.

File: src/containers/Application/ApplicationHome.tsx
Comment (andreievg) [inline src/containers/Application/ApplicationHome.tsx:34]: this should be replace, otherwise back button will always jump back to home then to summary

File: src/containers/Application/ApplicationHome.tsx
Comment (andreievg) [reply src/containers/Application/ApplicationHome.tsx:47]: I agree

**Review by andreievg — COMMENTED** [review]
Looks pretty good, I couple of minor comments. Pending Carl's approval

**Comment by andreievg** [comment]
@nmadruga summary of outstanding changes:

https://github.com/openmsupply/application-manager-web-app/pull/375#discussion_r583218514
https://github.com/openmsupply/application-manager-web-app/pull/375#discussion_r583213911

File: src/containers/Application/ApplicationHome.tsx
Comment (nmadruga) [reply src/containers/Application/ApplicationHome.tsx:47]: Yep

Comment (nmadruga) [reply src/containers/Application/ApplicationHome.tsx]: Done: 7068dd3

File: src/containers/Application/ApplicationHome.tsx
Comment (CarlosNZ) [reply src/containers/Application/ApplicationHome.tsx:34]: I found that "replace" made going back impossible, since the previous URL was wiped out (or maybe I was just using it wrong).

Perhaps the Router `.go(2)` method might be required, or perhaps play round with options until we're sure we get the intended behaviour. (https://reactrouter.com/web/api/history)

File: src/components/Application/ApplicationHeader.tsx
Comment (nmadruga) [reply src/components/Application/ApplicationHeader.tsx:36]: Another one to have a look as part of [Epic#10 - App layout](https://github.com/openmsupply/application-manager-web-app/issues/78) ?

File: src/containers/Application/ApplicationHome.tsx
Comment (nmadruga) [reply src/containers/Application/ApplicationHome.tsx:34]: Cool. Thanks changing this now.

File: src/components/Sections/SectionsProgress.tsx
Comment (nmadruga) [reply src/components/Sections/SectionsProgress.tsx]: Done: a6d10ce

File: src/containers/Application/ApplicationHome.tsx
Comment (nmadruga) [reply src/containers/Application/ApplicationHome.tsx:34]: Leaving this as `push` for now. If another issue is required to decide and make this change I think I can be added to the [Epic#10](https://github.com/openmsupply/application-manager-web-app/issues/78)

**Comment by andreievg** [comment]
@CarlosNZ pending your re-review

File: src/components/Sections/SectionsProgress.tsx
Comment (CarlosNZ) [inline src/components/Sections/SectionsProgress.tsx:56]: Can you also add the `as={Link}` property here?

**Review by CarlosNZ — CHANGES_REQUESTED** [review]
Couple of tiny things.

Also -- is it just me, or are you getting a compile error due to the dropdownChoice/ApplicationView plugin?

**Comment by nmadruga** [comment]
> Also -- is it just me, or are you getting a compile error due to the dropdownChoice/ApplicationView plugin?

No. I'm getting it too, which is annoying since it doesn't hot reload the app correctly after changes. I saw where the problem is but don't know how to fix... And it seems to work fine, so it's just playing with the error message there to fix.

File: src/containers/Application/ApplicationHome.tsx
Comment (nmadruga) [reply src/containers/Application/ApplicationHome.tsx:62]: Yep, good idea

File: src/components/Sections/SectionsProgress.tsx
Comment (nmadruga) [reply src/components/Sections/SectionsProgress.tsx:56]: Sure

**Comment by CarlosNZ** [comment]
> > Also -- is it just me, or are you getting a compile error due to the dropdownChoice/ApplicationView plugin?
> 
> No. I'm getting it too, which is annoying since it doesn't hot reload the app correctly after changes. I saw where the problem is but don't know how to fix... And it seems to work fine, so it's just playing with the error message there to fix.

It's working okay in the branch I'm currently working on, so I think some of the type definitions for ApplciationViewProps must have changed somewhere. I'll have a quick look

**Comment by nmadruga** [comment]
Cool. Thanks for having a look on this. You can commit to this PR once you find and then we can probably merge it.

**Comment by CarlosNZ** [comment]
> Cool. Thanks for having a look on this. You can commit to this PR once you find and then we can probably merge it.

Yeah, the error is on `master` too, so its nothing to do with the changes in here. Will hopefully be an easy fix then we can merge this.

**Comment by CarlosNZ** [comment]
Screen Shot 2021-03-03 at 5 44 46 PM
Here's the culprit -- the `error` prop on Dropdown only has boolean values not the more specific object. Will fix now.

**Review by CarlosNZ — APPROVED** [review]
LGTM now!

**Pull request merged**
